src/1-theory/3-input-and-output.adoc

```diff
@@ -0,0 +1,593 @@
+* 410 - Data has been deleted, deactivated, suspended, etc.
+* 415 - The request had a `Content-Type` which the server does not know
+how to handle
+* 429 - Rate Limited, which means take a breather, sleep a bit, try
+again
+* 500 - Something unexpected happened, and it is the APIs fault
+* 503 - API is not here right now, please try again later
+
+You might spot others popping up from time to time, so check on
+http://http.cat/[http.cats] (or
+http://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml[iana.org]
+for a more formal list) when you see one that's not familiar.
+
+== Mime Types
+
+HTTP APIs can work with all sorts of data. Whilst SOAP may have been restricted
+to XML, REST and GraphQL can work with any response types. gRPC is kinda stuck
+with Protobuf.
+
+An API can support almost unlimited options, but of course building support for
+every content type ever would be a rather laborious job. There are a few we can
+cut out early on.
+
+=== Ditch Form Data
+
+"Form Data" uses the `application/x-www-form-urlencoded` mime type, and mostly
+only seems to be used by PHP developers. Luckily most other folks ignore this
+wholeheartedly.
+
+.Form data doesn't really have data types, just awkward strings.
+[source]
+----
+foo=something&bar=1&baz=0
+----
+
+Another issue with form data is similar to how XML suffers a lack of obvious
+data types. For example, to handle a boolean a client has to send `1` or `0`,
+because `bar=true` would be `string("true")` on the server-side.
+
+Data types are important, so let's not just throw them out the window for the
+sake of "easy access to our data", especially as most web application frameworks
+have something like `$request->body->foo` to easily get to the foo property.
+
+WARNING: Rails is awful at this. If you have a `?foo=a` query string parameter,
+and you also send `{ "foo": "b" }` in the HTTP body, then `params[:foo]` will be
+set to `"b"` as the latter overrides the former. Code that you build, including
+any generic frameworks/tooling that you release, should avoid conflating query
+strings and body properties at all costs. They're different things and this
+nonsense causes confusion, especially when you realize that `params[:action]`
+means "controller method name" and actually overrides whatever is in `?action=`
+_without_ anything being in the body...
+
+.Instead of form data, use a nice JSON object.
+[source,http]
+----
+POST /checkins HTTP/1.1
+Host: api.example.org
+Content-Type: application/json
+
+{
+    "checkin": {
+        "place_id" : 1,
+        "message": "This is a bunch of text.",
+        "with_friends": [1, 2, 3, 4, 5]
+    }
+}
+----
+
+This is a perfectly valid HTTP body for a checkin. You know what they are
+saying. You know who the user is from their auth token. You know who they are
+with and you get the benefit of having it wrapped up in a single `checkin` key
+for simple documentation, and, easy "You sent a checkin object to the user
```

Comment:
```suggestion
and, easy "You sent a checkin object to the user
```
I think I know what you're trying to say here, but perhaps it could be said in its own sentence, unrelated to the first part? Looks like a thought shortcut a bit. :)